### ISSUE TITLE
Support yarn v2

### ISSUE THREAD

**Issue by oliversalzburg** [microsoft/vscode-vsce#517]
In a repository using yarn 2, `vsce package` will fail trying to enumerate dependencies:

```shell
 ERROR  Command failed: yarn list --prod --json
```

To reproduce, in an extension repository, use:

```shell
yarn set version berry
yarn install
vsce package
```

**Comment by joaomoreno** [comment]
What does `yarn list --prod --json` return?

Can I clone your repo and try it out?

**Comment by oliversalzburg** [comment]
It returns `Usage Error: Couldn't find a script named "list".`, as it's not a valid yarn2 command: https://yarnpkg.com/cli/

My reproduction should work in any extension repository, but I created a fresh one that should showcase it perfectly: https://github.com/oliversalzburg/vsce-yarn2-support

**Comment by SheepReaper** [comment]
Likewise, yarn 2 omits the run command since it's not needed. Yarn 2 is not meant to be installed globally (yet) _If_ there is yarn available in global, then it's probably v1. I don't know if vscode is running an embedded yarn or global yarn, but ordinarily yarn v1 checks if yarn 2 (or any yarn) is installed in the repository and runs it instead.

In my case, running an npm script from the side panel results in this:
```
The terminal process "C:\Program Files\PowerShell\7\pwsh.exe -Command yarn run dev" terminated with exit code: 1.
```

Which is incorrect, it should just be `yarn dev`

**Comment by jamesinaxx** [comment]
I had the same issue, here was my temporary fix, although it can be annoying,
1. *Optional*: Checkout a new temp branch so that package-lock.json and other npm files to carry over
2. Run `npm i`
3. Run `vsce package --no-yarn`

**Comment by latipun7** [comment]
just info, as of now `yarn list` still not implemented in yarn berry: https://yarnpkg.com/getting-started/migration#not-implemented-yet

**Comment by SchoofsKelvin** [comment]
Not the cleanest/best solution, but this is how I circumvented this issue:
- Configure your `.yarnrc.yml` with a `lockfileFilename` different from `yarn.lock`
- Run `vsce` with the `--yarn` parameter (since it won't detect Yarn using the `yarn.lock` anymore)

This will make `vsce` skip its whole process of looking at your dependencies and partially including `node_modules` in the packaged extension. Skipping this process also skips its internal `yarn list` _(and because of `--yarn` also `npm ls`)_ which prevents this error from happening.

The only issue with this is that `vsce` might bundle some items in `node_modules` if it isn't in `.vscodeignore` or something similar. You'd lose that "bundling" of course (which you would lose anyway if using Yarn PnP), but if your extension doesn't have any dependencies or you bundle it using e.g. Webpack, this is a non-issue.

Of course, having `vsce` support Yarn 2 natively is still better.

**Comment by itspngu** [comment]
With the freeze of yarn v1 this should perhaps be re-prioritized.

**Comment by YXL76** [comment]
> With the freeze of yarn v1 this should perhaps be re-prioritized.

See #439. I think yarn v2 is supported.

**Comment by ssbarnea** [comment]
This ticket is already two years old and we still do not have full support for yarn 2.

For example `vsce ls` command is fully broken because:

```
$ npx vsce ls
 INFO  Detected presence of yarn.lock. Using 'yarn' instead of 'npm' (to override this pass '--no-yarn' on the command line).
 ERROR  Command failed: yarn list --prod --json
```

There is no `list` in yarn 2 and if you try to pass `--no-yarn` to vsce, you endup with the other problems and failures.

I guess that for the moment I will need to replace `vsce ls` with `zipinfo -1 *.vsix`.

**Comment by oliversalzburg** [comment]
In general, it seems like very poor practice to rely on a specific third party shell application to perform anything of meaning in a NodeJS project, _especially_ when that third party application is a package manager.

Which is why I have given up on vsce a long time ago. Please open a new issue to track the total absence of progress on this issue, if you need to.

**Comment by joaomoreno** [comment]
Contributions here are more than welcome! Yarn v2 just isn't that big of a deal for us and most of our users (only 8 upvotes here), so this issue never makes it to the front page. Also, [and I'm sure you know](https://code.visualstudio.com/updates), we're constantly busy with plenty of other work.

**Comment by oliversalzburg** [comment]
If you need to keep track of your lack of interest to work on this subject, please create your own internal ticket. I don't want to be bothered by activity on this repository.

**Comment by joaomoreno** [comment]
@oliversalzburg, please unsubscribe.

**Comment by SheepReaper** [comment]
Did you all see the comment about reacting to the OP? That's how they track heat. I wish that was more obvious.

**Comment by arendjr** [comment]
I also ran into this issue, but resolved it by creating a Yarn plugin that implements the `list` command. I have only tested this on our own repository, so I may not have covered all edge cases, but it works for us.

See: https://gist.github.com/arendjr/415747652a8c79f12b005ce3cfb2f808

If there is interest in contributing this to Yarn itself, please let me know.

**Comment by lordrip** [comment]
Hi @arendjr, thanks for sharing your plugin. I think that it's a good idea of having your plugin listed on the `yarn` website.

How can I help to make this? I guess that we would need to put the plugin in a dedicated package perhaps.

**Comment by lordrip** [comment]
Hi @joaomoreno, from your point of view, how would you envision this issue moving forward? I might be lacking context but I think that having @arendjr plugin installed on `yarn berry` repositories it's good and probably it will evolve to support the previous use cases from `yarn list` from `yarn v1`, but at the same time, maybe `vscode-vsce` shouldn't depend on the users installing the plugin but rather have this functionality embedded.

Maybe another way to handle it could be listing @arendjr's plugin as a prerequisite of `vscode-vsce` for supporting `yarn berry`

**Comment by ivangabriele** [comment]
Thank you so much for working on this @lordrip :). Fyi, it seems `yarn list` has simply [be renamed/restructured](https://yarnpkg.com/getting-started/migration#renamed) to `yarn info --recursive` in case that help you with your WIP.

**Comment by lordrip** [comment]
Hi @ivangabriele I haven't worked on this yet, I would like to make sure that this is still needed first. Thanks for the heads-up about the renaming, I gave it a try and it seems that the JSON output is not exactly right:

Using:
```bash
yarn info --recursive --json
```

this is the output, notice how it will need a little bit of adjustment before using it:
```json
{
    "value":"@4tw/cypress-drag-drop@npm:2.2.4",
    "children":{
        "Instances":1,
        "Version":"2.2.4"
    }
}{
    "value":"@adobe/css-tools@npm:4.2.0",
    "children":{
        "Version":"4.2.0"
    }
}{
    "value":"@ampproject/remapping@npm:2.2.1",
    "children":{
        "Version":"2.2.1",
        "Dependencies":[
            {
                "descriptor":"@jridgewell/gen-mapping@npm:^0.3.0",
                "locator":"@jridgewell/gen-mapping@npm:0.3.3"
            },
            {
                "descriptor":"@jridgewell/trace-mapping@npm:^0.3.9",
                "locator":"@jridgewell/trace-mapping@npm:0.3.18"
            }
        ]
    }
}
```

**Comment by ivangabriele** [comment]
@lordrip Indeed I didn't check in detail how it's used yet. Normally I think that could be a one-shot fix by updating this [function](https://github.com/microsoft/vscode-vsce/blob/main/src/npm.ts#L157C7-L183).

I can at least confirm I had the same issue today on a brand new extension using Yarn stable (v3.6.1).

In the meantime I used [@arendjr script](https://github.com/microsoft/vscode-vsce/issues/517#issuecomment-1521712750).

I can try to poke a PR if you wish ^^.

**Comment by lordrip** [comment]
Sure, I'm not the maintainer but I'm confident that it will help a lot.

If you happen to submit a pull request I can lend you a hand if needed.

If it helps, I adjusted a bit @arendjr 's [plugin here](https://github.com/KaotoIO/vscode-kaoto/blob/1a9b80d73416bad6e5f588aa2bef30edd38ca76a/.yarn/plugins/plugin-list.js) to incorporate the `yarn link` workflow as well

**Comment by arendjr** [comment]
Would it help if I moved my script into its own repository?

**Comment by lordrip** [comment]
Hi @arendjr, as I mentioned in a few comments above, I think that there's value in trying to publish your script in the [`yarn plugin` list](https://github.com/yarnpkg/berry/blob/master/packages/gatsby/content/features/plugins.md).

Plenty of plugin examples exist, for instance, [yarn-plugin-env-vars](https://github.com/scinos/yarn-plugin-env-vars).

That being said, I don't think that both efforts are mutually exclusive, at least how I see it, it would be quicker to have a public plugin, installed through the CLI, like:

```bash
yarn add arendjr-plugin
```

while we wait for a dedicated implementation in this repository, which probably would be the best approach in the long run :smiley:

**Comment by arendjr** [comment]
Alrighty, this should do the trick :) https://github.com/arendjr/yarn-plugin-list

I have incorporated your support for Yarn resolutions as well, @lordrip

**Comment by ivangabriele** [comment]
Thanks again here. Fyi I will try to open a PR-fix today or tomorrow for this issue.

**Comment by anbnyc** [comment]
I was blocked by this issue for a while, but in case it is helpful to others, my workaround was to use esbuild to bundle, then pass `--no-dependencies` to my package command. This skips the outdated `yarn list` command. Credit for this idea: https://github.com/microsoft/vscode-vsce/issues/421#issuecomment-1038911725

For emphasis:
> there is no need to include npm dependencies when package and publish, all things is in your bundle

As an additional aside, it looks like the compatibility issue looks like it was introduced in `v2.9.2`: https://github.com/microsoft/vscode-vsce/releases/tag/v2.9.2. If you are OK using an earlier version of `vsce`, this might help too.

**Comment by joaomoreno** [comment]
Closing this as out of scope. As part of our recurring maintenance milestone, we've decided to limit the amount of supported package managers to npm and yarn v1 only.

That being said, I'm glad workarounds exist for anyone who is blocked.

**Comment by BasixKOR** [comment]
Have you considered using an off-the-shelf solution for resolving Node.js dependencies instead? It seems like the package manager-dependent code is just used for listing which `node_modules` should be put into the VSIX package, and it has been solved by the community countless times like https://github.com/vercel/nft or any sufficient Node.js resolver implementation.

I think resolving dependencies from the code would lead to better accuracy (since you cannot be so sure all the package maintainers are managing `devDependencies` and `dependencies` well) and better interoperability overall without juggling over every package manager VSCE may need to support.

**Comment by yangwooseong** [comment]
How can I avaoid this error when I want to include dependencies?

**Comment by arendjr** [comment]
@yangwooseong I believe the plugin I created should work if you have dependencies: https://github.com/arendjr/yarn-plugin-list

But I haven't used or tested it for quite a while now. I'd be happy to hear if it still works (or not).

**Comment by yangwooseong** [comment]
@arendjr Thanks for your plugin!

My repository is a monorepo structured like this:

```
packages
  - vscode-extension
  - ...other packages
yarn.lock
```

To make it work with the plugin, I modified the line that reads the yarn.lock file to `fs.readFileSync("../../yarn.lock")`. Even though I did this and can see the VS Code-related dependencies with `yarn list --prod --json`, they don’t appear when I run `vsce ls` as well as `vsce pakcage`.
